### PR TITLE
[#446][Part3] Add OpenAPI to support creating API documentation with OpenAPI

### LINKED ISSUE #446: Add OpenAPI to support creating API documentation with OpenAPI
## Why

The team now has [OpenAPI Training Path](https://notion.so/d76224c3480242aeb4e8c71055a76398?pvs=25) to learn and practice. [OpenAPI on Compass](https://github.com/nimblehq/compass/pull/1283) for developers to follow conventions and best practices. NO OpenAPI boilerplates currently exist in the rails templates to support web developers in creating API documentation with OpenAPI.

## Acceptance Criteria

- Add OpenAPI addon as a required addon in the template.
- Add OpenAPI tooling (Spectral, Elements, Lint).
- Create an endpoint (e.g., `GET {{base_url}}/api/docs/openapi`) to access the API documentation through the browser.
- Add Mock Server with Prism (related to this [API Mock Server with OpenAPI](https://www.notion.so/API-Mock-Server-with-OpenAPI-83fda1653eb1409fab9b08c1fcf466fd?pvs=21)).
- Add CI/CD workflow to run linting and deploy mock server to fly.io.

## Resources

- [Add OpenAPI to Web templates](https://www.notion.so/nimblehq/Add-OpenAPI-to-Web-templates-5b843ba050464edb80d14dfbad6da0fe?pvs=4)
- https://github.com/nimblehq/api-mock-templates/pull/3
- https://github.com/nimblehq/nimble-survey-web

## Who Benefits?

Everyone

### PR DESCRIPTION
**Pull request by sanG-github**
close #446

## What happened 👀

- Added `Dockerfile.mock`
- Added `fly.toml`
- Added `deploy_mock_server.yml`

## Insight 📝

Support `mock_server` add-on for the templates. This will help developers to deploy a mock server based on OpenAPI spec.

Here is the checklist of #446 when this pull request is merged:

- [x] Add OpenAPI addon as an mandatory addon in the template.
- [x] Add OpenAPI tooling (Spectral, Lint).
- [x] Create an endpoint (e.g., GET {{base_url}}/api/docs/openapi) to access the API documentation through the browser.
- [x] Add Mock Server with Prims (related to this [API Mock Server with OpenAPI](https://www.notion.so/API-Mock-Server-with-OpenAPI-83fda1653eb1409fab9b08c1fcf466fd?pvs=21)).
- [x] Add CI/CD workflow to run linting and deploy mock server to fly.io:
  - [x] Linting
  -	[x] Deploy Mock Server

## Proof Of Work 📹

| Deployed successfully on local.                                                                                                        	|
|----------------------------------------------------------------------------------------------------------------------------------------	|
| Screenshot 2023-09-07 at 16 13 05 	|

### CODE REVIEW

**Review by olivierobert — APPROVED** [review]

**Review by hoangmirs — COMMENTED** [review]
As this PR belongs to the OpenAPI addon, should we move all template files into that folder instead? And find a way to integrate it with other add-ons (Docker, GitHub)?

**Comment by malparty** [comment]
After discussion with Sang, OpenAPI might NOT be an add-on, but a "mandatory" part of the template, included in any of the variants (API & Web). I'll wait until later this week to confirm that with Su V. If it's the case, I would argue that it will help keep the template code simpler and easier to test :D

**Comment by olivierobert** [comment]
@malparty @sanG-github  It cannot be a compulsory add-on if no API is built. It is only. You can refer to the Gin templates: 

https://github.com/nimblehq/gin-templates/blob/ad4b942bf841d834665d7198ecbba810eebfcd86/hooks/pre_gen_project.py#L36-L40

**Comment by malparty** [comment]
All right, then let's implement it so that it does not leak in a new project that does not select the add-on 🙇

**Review by suho — COMMENTED** [review]
For a mock server, it should be an add-on (you can check [gin-templates](https://github.com/nimblehq/gin-templates/pull/108/files#diff-c2067af38dcc45bd90a9c6b3ef58f6995ac2e7386dd61e802911b3deff05908eR7))

- For API variant, we should show a message to ask if the developer want to have a mock server
- For Web variant, if the developer add openapi add on, then we will ask to have a mock server

**Review by suho — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 openapi_addon_prompt = Y\n
 
+# Y - in response to Would you like to add the Mock Server addon?
 # Y - in response to Would you like to add the Github addon?
 # Y - in response to Would you like to add the SemaphoreCI addon?
 # Y - in response to Would you like to add the Nginx addon?
 # Y - in response to Would you like to add the Phrase addon?
 # Y - in response to Would you like to add the Devise addon?
-common_addon_prompts = Y\nY\nY\nY\nY\n
+common_addon_prompts = Y\nY\nY\nY\nY\nY\n
```
Comment (malparty) [inline Makefile:10]: Rebase needed 😇

File: Makefile [hunk 2]
```diff
@@ -9,16 +10,17 @@
 # Y - in response to Would you like to add the Hotwire addon?
 
 create_web:
-	printf "Y\nY\nY\nY\nY\nY\nY\nY\nY\n" | rails new $(APP_NAME) -m ./template.rb -T ${OPTIONS}
+	printf "Y\nY\nY\nY\nY\nY\nY\nY\nY\nY\n" | rails new $(APP_NAME) -m ./template.rb -T ${OPTIONS}
 
 # Y - in response to Would you like to add the Github addon?
+# Y - in response to Would you like to add the Mock Server addon?
 # Y - in response to Would you like to add the SemaphoreCI addon?
```
Comment (malparty) [inline Makefile:17]:
```suggestion
# Y - in response to Would you like to add the SemaphoreCI addon?
# Y - in response to Would you like to add the Mock Server addon?
```

Comment (malparty) [reply Makefile:17]: Sorry, my eyes where looking at the web part, 


image

so when I saw 

image

I thought the ordering was wrong :harold:

File: .template/spec/addons/base/docker/template_spec.rb [hunk 3]
```diff
@@ -5,6 +5,10 @@
     expect(file('Dockerfile')).to exist
   end
 
+  it 'creates the Dockerfile.mock' do
+    expect(file('Dockerfile.mock')).to exist
+  end
+
```
Comment (malparty) [inline .template/spec/addons/base/docker/template_spec.rb:11]: Let's move this spec inside the OpenAPI tests :white_check_mark:

Comment (sanG-github) [reply .template/spec/addons/base/docker/template_spec.rb]: I thought I moved it to OpenAPI 🤯 
I moved it in 7a73e42 ✅

**Review by malparty — APPROVED** [review]

**Comment by malparty** [comment]
@sanG-github last rebase needed :harold:

**Pull request merged**